src/dialer/index.js

```diff
@@ -0,0 +1,129 @@
+'use strict'
+
+const multiaddr = require('multiaddr')
+const errCode = require('err-code')
+const AbortController = require('abort-controller')
+const anySignal = require('any-signal')
+const debug = require('debug')
+const log = debug('libp2p:dialer')
+log.error = debug('libp2p:dialer:error')
+const { DialRequest } = require('./dial-request')
+
+const { codes } = require('../errors')
+const {
+  DIAL_TIMEOUT,
+  MAX_PARALLEL_DIALS,
+  MAX_PER_PEER_DIALS
+} = require('../constants')
+
+class Dialer {
+  /**
+   * @constructor
+   * @param {object} options
+   * @param {TransportManager} options.transportManager
+   * @param {Peerstore} peerStore
+   * @param {number} options.concurrency Number of max concurrent dials. Defaults to `MAX_PARALLEL_DIALS`
+   * @param {number} options.timeout How long a dial attempt is allowed to take. Defaults to `DIAL_TIMEOUT`
+   */
+  constructor ({
+    transportManager,
+    peerStore,
+    concurrency = MAX_PARALLEL_DIALS,
+    timeout = DIAL_TIMEOUT,
+    perPeerLimit = MAX_PER_PEER_DIALS
+  }) {
+    this.transportManager = transportManager
+    this.peerStore = peerStore
+    this.concurrency = concurrency
+    this.timeout = timeout
+    this.perPeerLimit = perPeerLimit
+    this.tokens = [...new Array(concurrency)].map((_, index) => index)
+    this.pendingDials = new Set()
+  }
+
+  /**
+   * Connects to the first success of a given list of `Multiaddr`. `addrs` should
+   * include the id of the peer being dialed, it will be used for encryption verification.
+   *
+   * @param {Array<Multiaddr>|Multiaddr} addrs
+   * @param {object} [options]
+   * @param {AbortSignal} [options.signal] An AbortController signal
+   * @returns {Promise<Connection>}
+   */
+  async connectToMultiaddr (addrs, options = {}) {
+    if (!Array.isArray(addrs)) addrs = [multiaddr(addrs)]
+
+    const dialAction = (addr, options) => {
+      if (options.signal.aborted) throw errCode(new Error('already aborted'), 'ERR_ALREADY_ABORTED')
+      return this.transportManager.dial(addr, options)
+    }
+    const dialRequest = new DialRequest({
+      addrs,
+      dialAction,
+      dialer: this
+    })
+
+    // Combine the timeout signal and options.signal, if provided
+    const timeoutController = new AbortController()
+    const signals = [timeoutController.signal]
+    options.signal && signals.push(options.signal)
+    const signal = anySignal(signals)
+    const timeoutId = setTimeout(() => timeoutController.abort(), this.timeout)
```

Comment:
I know this doesn't really save any LoC but it just makes it a bit more of a concrete thing that people might reach for instead of rolling their own timeouts. I'll need this in js-ipfs too:

```js
class TimeoutController extends AbortController {
  constructor (ms) {
    super()
    this._timeoutId = setTimeout(() => this.abort(), ms)
  }
  abort () {
    clearTimeout(this._timeoutId)
    return super.abort()
  }
  clear () {
    clearTimeout(this._timeoutId)
  }
}
```

...also, since this is the controller being used to abort dials when libp2p stops, this class actually clears the timeout if `abort` is called before it expires - currently these timeouts remain running and may cause `abort` to be called multiple times. I think that's not actually a problem but having timers still running might keep the process running (although you might not seeing this right now because `pendingDials` are deleted and all of the timers are probably all garbage collected).